File: rater.js
import React, { useState, useEffect } from "react";
import { Button } from "@/components/ui/button";

const items = [
  "Louis", "Ben", "Dan", "Alisha", "Zak", "Corin", "Cuish", "Morgan", "Perry", 
  "Sadie", "Leah", "Amy", "Charlie", "Harry", "Sean", "Ruby", "Tanith", "Reece", 
  "Eleanor", "Makoto", "Masami", "Meriel"
];

const shufflePairs = (arr) => {
  let pairs = [];
  for (let i = 0; i < arr.length; i++) {
    for (let j = i + 1; j < arr.length; j++) {
      pairs.push([arr[i], arr[j]]);
    }
  }
  return pairs.sort(() => Math.random() - 0.5);
};

export default function VotingApp() {
  const [pairs, setPairs] = useState(shufflePairs(items));
  const [pairIndex, setPairIndex] = useState(0);
  const [scores, setScores] = useState(() => {
    let initialScores = {};
    items.forEach((item) => (initialScores[item] = 0));
    return initialScores;
  });
  const [showResults, setShowResults] = useState(false);

  const vote = (winner) => {
    setScores((prevScores) => ({
      ...prevScores,
      [winner]: prevScores[winner] + 1,
    }));
    if (pairIndex + 1 < pairs.length) {
      setPairIndex(pairIndex + 1);
    } else {
      setShowResults(true);
    }
  };

  return (
    <div className="flex flex-col items-center p-6">
      {!showResults ? (
        <div className="text-center">
          <h2 className="text-xl font-bold mb-4">Which do you prefer?</h2>
          <div className="flex gap-4">
            <Button onClick={() => vote(pairs[pairIndex][0])} className="text-lg p-4">
              {pairs[pairIndex][0]}
            </Button>
            <Button onClick={() => vote(pairs[pairIndex][1])} className="text-lg p-4">
              {pairs[pairIndex][1]}
            </Button>
          </div>
        </div>
      ) : (
        <div className="text-center">
          <h2 className="text-xl font-bold mb-4">Popularity Ranking</h2>
          {Object.entries(scores)
            .sort((a, b) => b[1] - a[1])
            .map(([item, score], index) => (
              <p key={item} className="text-lg">{index + 1}. {item} (Score: {score})</p>
            ))}
        </div>
      )}
    </div>
  );
}

File: rater.py
from flask import Flask, render_template, request, redirect, url_for
from itertools import combinations
from collections import defaultdict
import random

app = Flask(__name__)

items = ["Louis", "Ben", "Dan", "Alisha", "Zak", "Corin", "Cuish", "Morgan", "Perry", "Sadie", "Leah", "Amy", "Charlie", "Harry", "Sean", "Ruby", "Tanith", "Reece", "Eleanor", "Makoto", "Masami", "Meriel"]


scores = defaultdict(int)
pairs = list(combinations(items, 2))
random.shuffle(pairs)
pair_index = 0

@app.route('/')
def index():
    global pair_index
    if pair_index < len(pairs):
        item1, item2 = pairs[pair_index]
        return render_template("vote.html", item1=item1, item2=item2)
    else:
        return redirect(url_for("results"))

@app.route('/vote', methods=['POST'])
def vote():
    global pair_index
    choice = request.form.get("choice")
    item1, item2 = pairs[pair_index]
    if choice == item1:
        scores[item1] += 1
    else:
        scores[item2] += 1
    
    pair_index += 1
    return redirect(url_for("index"))

@app.route('/results')
def results():
    sorted_items = sorted(items, key=lambda x: scores[x], reverse=True)
    return render_template("results.html", results=sorted_items, scores=scores)

if __name__ == '__main__':
    app.run(debug=True)
